src/helpers/messaging.rs

```diff
@@ -143,46 +121,80 @@ impl<N: Network> Mesh<'_, '_, N> {
+    /// 10 messages sent trigger a network request.
+    pub send_buffer_capacity: u32,
+}
+
 impl<N: Network> Gateway<N> {
-    pub fn new(role: Identity, network: N) -> Self {
+    pub fn new(role: Identity, network: N, config: GatewayConfig) -> Self {
         let (tx, mut receive_rx) = mpsc::channel::<ReceiveRequest>(1);
-        let mut message_stream = network.message_stream();
+        let (envelope_tx, mut envelope_rx) = mpsc::channel::<(ChannelId, MessageEnvelope)>(1);
+        let mut message_stream = network.stream();
+        let mut network_sink = network.sink();
+
+        let control_handle = tokio::spawn(async move {
+            // to make forward progress, we periodically check if the system is stalled
+            // if nothing happens for long period of time, we try to unblock it by flushing
+            // the data that remains inside buffers. Note that the interval picked here is somewhat
+            // random - waiting for too long will result in elevated latencies. On the other hand,
+            // sending buffers that are half-full will lead to underutilizing the network
+            const INTERVAL: Duration = Duration::from_millis(200);
+
+            let mut receive_buf = ReceiveBuffer::new();
+            let mut send_buf = SendBuffer::new(config.send_buffer_capacity);
 
-        tokio::spawn(async move {
-            let mut buf = HashMap::<ChannelId, MessageBuffer>::new();
+            let sleep = tokio::time::sleep(INTERVAL);
+            tokio::pin!(sleep);
 
             loop {
                 // Make a random choice what to process next:
                 // * Receive a message from another helper
                 // * Handle the request to receive a message from another helper
+                // * If send buffer is full, send it down
                 tokio::select! {
                     Some(receive_request) = receive_rx.recv() => {
                         tracing::trace!("new {:?}", receive_request);
-                        buf.entry(receive_request.channel_id)
-                           .or_default()
-                           .receive_request(receive_request.record_id, receive_request.sender);
+                        receive_buf.receive_request(receive_request.channel_id, receive_request.record_id, receive_request.sender);
                     }
                     Some((channel_id, messages)) = message_stream.next() => {
                         tracing::trace!("received {} message(s) from {:?}", messages.len(), channel_id);
-                        buf.entry(channel_id)
-                           .or_default()
-                           .receive_messages(messages);
+                        receive_buf.receive_messages(&channel_id, messages);
+                    }
+                    Some((channel_id, msg)) = envelope_rx.recv() => {
+                        if let Some(buf_to_send) = send_buf.push(channel_id.clone(), msg) {
```

Comment:
So I take it that reordering is a TBD?  From what I can see, you are sending things in arbitrary order on each channel.

Reply:
correct, `send_buf` is dumb and just accumulates up to `batch_size` records before sending them down to network layer. I hope we don't need to change this particular piece of code though because send buffers should take care of ordering.

There is another important piece that is missing here - window size update. We should probably slow down producers that went too far ahead by looking in `record_id` and blocking their `send` if `record_id` is outside of accepted range. I think we need some cooperation between send buffers and the gateway to make this happen.

I will add a TODO so we don't forget about it